Amazon.Ion.ObjectMapper/IonObjectSerializer.cs

```diff
@@ -124,6 +120,118 @@ public override void Serialize(IIonWriter writer, object item)
             writer.StepOut();
         }
 
+        private bool DeserializeProperty(PropertyInfo property, IIonReader reader, IonType ionType, ref object deserialized)
+        {
+            if (IsReadOnlyProperty(property))
+            {
+                // property.SetValue() does not work with a readonly property.
+                // logic for handling deserializing readonly properties happens during field processing
+                // when we detect backing fields for the property.
+                return false;
+            }
+
+            deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
+            if (options.IgnoreDefaults && deserialized == default)
+            {
+                return false;
+            }
+
+            return true;
+        }
+        
+        private bool DeserializeField(FieldInfo field, IIonReader reader, IonType ionType, ref object deserialized)
+        {
+            deserialized = ionSerializer.Deserialize(reader, field.FieldType, ionType);
+            
+            if (options.IgnoreReadOnlyFields && field.IsInitOnly)
+            {
+                return false;
+            }
+            
+            if (options.IgnoreDefaults && deserialized == default)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private object DeserializeWithIonConstructor(ConstructorInfo ionConstructor, IIonReader reader)
+        {
+            var parameters = ionConstructor.GetParameters();
+
+            // Compute mapping between parameter names and index in parameter array so we can figure out the
+            // correct order of the constructor arguments.
+            var paramIndexMap = new Dictionary<string, int>();
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                var ionPropertyName = (IonPropertyName) parameters[i].GetCustomAttribute(typeof(IonPropertyName));
+                if (ionPropertyName == null)
+                {
+                    throw new NotSupportedException(
+                        $"Parameter '{parameters[i].Name}' is not specified with the [IonPropertyName] attribute " +
+                        $"for {targetType.Name}'s IonConstructor. All constructor arguments must be annotated " +
+                        "so we know which parameters to set at construction time.");
+                }
+
+                paramIndexMap.Add(ionPropertyName.Name, i);
+            }
+
+            reader.StepIn();
+
+            // Deserialize Ion and organize deserialized results into three categories:
+            // 1. Values to be passed into the Ion constructor.
+            // 2. Properties to be set after construction.
```

Comment:
What about methods that are annotated?

Reply:
That relies on implementation in another PR. I will add that logic once the other PR is merged in.

In the meantime I will downgrade this PR to a draft PR.